tiledb/sm/misc/parallel_functions.h

```diff
@@ -180,26 +180,67 @@ Status parallel_for(
 
   assert(tp);
 
+  /*
+   * Mutex protects atomicity of `failed*` local variables together. The first
+   * subrange to fail determines the return or exception value.
+   */
+  std::mutex failed_mutex;
+  /*
+   * If we were checking this variable inside either the main loop or the one in
+   * `execute_subrange`, then it would be better to use `atomic_bool` to lessen
+   * the lock overhead on the mutex. As it is, we do not prematurely stop any
+   * loop that is not the first to fail.
+   */
   bool failed = false;
-  Status return_st = Status::Ok();
-  std::mutex return_st_mutex;
-
-  // Executes subrange [subrange_start, subrange_end) that exists
-  // within the range [begin, end).
+  optional<std::exception_ptr> failed_exception{nullopt};
+  optional<Status> failed_status{nullopt};
+
+  /*
+   * Executes subrange [subrange_start, subrange_end) that exists within the
+   * range [begin, end).
+   *
+   * If all the functions `F` return OK, then so does this function. If a
+   * function fails but is not the first do so in this `parallel_for`
+   * invocation, then this function returns OK. If a first function to fail
+   * returns not OK, then this function returns that value. If a first function
+   * to fail throws, then this function throws that value.
+   */
   std::function<Status(uint64_t, uint64_t)> execute_subrange =
-      [&failed, &return_st, &return_st_mutex, &F](
+      [&failed, &failed_exception, &failed_status, &failed_mutex, &F](
           const uint64_t subrange_start,
           const uint64_t subrange_end) -> Status {
     for (uint64_t i = subrange_start; i < subrange_end; ++i) {
-      const Status st = F(i);
-      if (!st.ok() && !failed) {
-        failed = true;
-        std::lock_guard<std::mutex> lock(return_st_mutex);
-        return_st = st;
+      Status st;
+      try {
+        st = F(i);
+        if (st.ok()) {
+          continue;
+        }
+        std::lock_guard<std::mutex> lock(failed_mutex);
+        if (!failed) {
+          failed_status = st;
+          failed = true;
+          return st;
+        }
+      } catch (...) {
+        std::lock_guard<std::mutex> lock(failed_mutex);
+        if (!failed) {
+          auto ce{std::current_exception()};
+          failed_exception = ce;
+          failed = true;
+          std::rethrow_exception(ce);
```

Comment:
The rethrown exception will then be caught by the `ThreadPool`.  Do we (eventually) want to rethrow all exceptions rather than just the first one caught?  (This would assume certain policies and use the return value of `wait_all`).  That would obviate the need to `catch` here.  We could even throw on a not-okay status.  But, I understand the need for backward-compatibility.

We should also (eventually) attach status information to our thrown exceptions because right now the assumption in `ThreadPool` is that exceptions come from outside of core (and that core always just returns statuses) -- but that is another discussion.

Reply:
I decided to have it throw in the subrange to mirror the behavior of returning a status in the same first-error situation. It won't be necessary if we have a thread-pool policy that completely covers our error returns, but our current one does not.